### ISSUE TITLE
Issue with Globber#regexp and the new Rails 4.2.5.1 glob pattern

### ISSUE THREAD

**Issue by tovodeverett** [fakefs/fakefs#328]
Rails 4.2.5.1 introduces some seriously evil glob patterns like this: `views/photos/image{.{en},}{.{html},}{+{},}{.{erb,builder,raw,ruby,coffee,haml},}`.  This totally breaks `FakeFS::Globber#regexp` in two ways.  It breaks because it has a `+` in the pattern, and it breaks because it has nested curly braces.

The workaround that includes both the fix for this issue (soon to be posted with a test) and for https://github.com/defunkt/fakefs/issues/326 is:
```
module FakeFS
  module Globber
    def regexp(pattern)
      regex_body = pattern.gsub('.', '\.')
                          .gsub('+') { '\+' }
                          .gsub('?', '.')
                          .gsub('*', '.*')
                          .gsub('(', '\(')
                          .gsub(')', '\)')

      curly_braces_re = %r{
                          (?<re>
                            \{
                              (?:
                                (?> [^{}]+ )
                                |
                                \g<re>
                              )*
                            \}
                          )
                        }x
      while regex_body =~ curly_braces_re
        regex_body = regex_body.gsub(curly_braces_re) do
          "(#{Regexp.last_match[1][1..-2].gsub(',', '|')})"
        end
      end

      regex_body = regex_body.gsub(/\A\./, '(?!\.).')

      /\A#{regex_body}\Z/
    end

    [:expand, :path_components, :regexp].each do |method_sym|
      old_method = instance_method(method_sym)
      define_method(method_sym) do |pattern|
        old_method.bind(self).(pattern.to_s)
      end
    end
  end
end
```

**Comment by tovodeverett** [comment]
Here's a cleaner version of that loop to repeatedly apply the gsub:

```
module FakeFS
  module Globber
    def regexp(pattern)
      regex_body = pattern.gsub('.', '\.')
                          .gsub('+') { '\+' }
                          .gsub('?', '.')
                          .gsub('*', '.*')
                          .gsub('(', '\(')
                          .gsub(')', '\)')

      curly_braces_re = %r{
                          (?<re>
                            \{
                              (?:
                                (?> [^{}]+ )
                                |
                                \g<re>
                              )*
                            \}
                          )
                        }x
      loop do
        break unless regex_body.gsub!(curly_braces_re) do
          "(#{Regexp.last_match[1][1..-2].gsub(',', '|')})"
        end
      end

      regex_body = regex_body.gsub(/\A\./, '(?!\.).')

      /\A#{regex_body}\Z/
    end

    [:expand, :path_components, :regexp].each do |method_sym|
      old_method = instance_method(method_sym)
      define_method(method_sym) do |pattern|
        old_method.bind(self).(pattern.to_s)
      end
    end
  end
end
```

**Comment by tovodeverett** [comment]
See https://github.com/fakefs/fakefs/pull/330 for a resolution.

**Comment by tovodeverett** [comment]
Resolved by merge of #336 in 0.9.0.
